Q# manual (merger of Q# P. Language and Q. Dev. Techniques) (Branch, not fork) (#486)
MicrosoftDocs/quantum-docs-pr: articles/manual/statements.md

---
title: Q# statements | Microsoft Docs 
description: Q# statements, Control Flow and other constructs
author: QuantumWriter
uid: microsoft.quantum.manual.statements
ms.author: Alan.Geller@microsoft.com 
ms.date: 12/11/2017
ms.topic: article
---

> # Statements, Control Flow and other constructs 

Reviewer: The capitalization here should be consistent.
```suggestion
# Statements, Control Flow, and Other Constructs 
```

> This section contains description of statements,work flow elements like loops or conditonals and other kind of constructs.

Reviewer: "workflow elements" reads a bit awkwardly to me, but I'm not sure what to suggest instead, sorry.

Author: How about "control flow structures"?

An immutable binding consists of the keyword `let`, followed by
a symbol or symbol tuple, an equals sign `=`, an expression to bind the symbol(s) to, and a terminating semicolon.
The type of the bound symbol(s) is inferred based on the expression on the right hand side. For instance:

```qsharp

> let measurementOperator = [PauliX, PauliZ, PauliZ, PauliX, PauliI];

Reviewer: ```suggestion
let measurementOperators = [PauliX, PauliZ, PauliZ, PauliX, PauliI];
```

```

> This assigns a particular array of Pauli operators to a variable called `measurementOperator`.

Reviewer: ```suggestion
This assigns a particular array of Pauli operators to an immutable variable called `measurementOperators`.
```

Note that none of these statements are allowed within the body of a function.
They are only valid within operations.

> ### Allocating Qubits

Reviewer: ```suggestion

### Allocating Qubits
```

If there aren't enough such qubits, then it will allocate new qubits
to complete the request.

### Intrinsic Operations
Once allocated, a qubit can then be passed to functions and operations.
In some sense, this is all that a Q# program can do with a qubit, as the actions that can be taken are all defined as operations.
More information of these operations can be found in [Intrinsic Operations and Functions](xref:microsoft.quantum.libraries.standard.prelude) and [Operations and Functions](xref:microsoft.quantum.manual.operations-functions).

Here we present an example of operation implementation:

> First, the single-qubit Pauli operators $X$, $Y$, and $Z$ are represented in Q# by the intrinsic operations `X`, `Y`, and `Z`, each of which has type `(Qubit => Unit is Adj + Ctl)`.

Reviewer: It might be good to link to the API references for `X`, `Y`, and `Z` here.

Author: Ok. Done in 6d27732ae2a8bf026a01f135d6fd4e7891328e2c

CNOT(control, target);
Adjoint T(q);
```
## Measurements ##

Using the `Measure` operation, which is a built in intrinsic non-unitary operation, we can extract classical information from an object of type `Qubit` and assign a classical value as a result, which has a reserved type `Result`, indicating that the result is no longer a quantum state. 
The input to `Measure` is a Pauli axis on the Bloch sphere, represented by an object of type `Pauli` (i.e., for instance `PauliX`) and an object of type `Qubit`. 

A simple example is the following operation which creates one qubit in the $\ket{0}$ state, then applies a Hadamard gate ``H`` to it and then measures the result in the `PauliZ` basis. 

```qsharp

> operation MeasurementOneQubit () : Result {

Reviewer: ```suggestion
operation MeasureOneQubit() : Result {
```

    // The following using block creates a fresh qubit and initializes it 
    // in the |0〉 state.
    using (qubit = Qubit()) {
        // We apply a Hadamard operation H to the state, thereby creating the 
        // state 1/sqrt(2)(|0〉+|1〉). 
        H(qubit); 
        // Now we measure the qubit in Z-basis.
        let result = M(qubit);
        // As the qubit is now in an eigenstate of the measurement operator, 
        // we reset the qubit before releasing it. 
        if (result == One) { X(qubit); }   
        // Finally, we return the result of the measurement. 
        return result;
    }
}
```

> 

Reviewer: It may be good to note that this can be done in a simpler fashion by using the libraries:

```Q#
open Microsoft.Quantum.Measurement;

operation MeasureOneQubit() : Result {
   // The following using block creates a fresh qubit and initializes it 
    // in the |0〉 state.
    using (qubit = Qubit()) {
        // We apply a Hadamard operation H to the state, thereby creating the 
        // state 1/sqrt(2)(|0〉+|1〉). 
        H(qubit); 
        // Now we measure the qubit in Z-basis and return the result.
        return MResetZ(qubit);
    }
}

Author: Cool, added in 20241a3d46fc38cf0db98e5aabffbddea3601ac0 and beb92d72a29afc12a2b8177985e0c00526a9be9a

> A slightly more complicated example is given by the following operation which returns the Boolean value `true` if all qubits in a register of type `Qubit[]` are in the state zero, when measured in a specified Pauli basis and `false` otherwise. 

Reviewer: ```suggestion
A slightly more complicated example is given by the following operation which returns the Boolean value `true` if all qubits in a register of type `Qubit[]` are in the state zero when measured in a specified Pauli basis, and `false` otherwise. 
```

```qsharp

> operation AllMeasurementsZero (qs : Qubit[], pauli : Pauli) : Bool {

Reviewer: In keeping with the style guide, it would be good if `qs` was the second argument.

Author: Done in d84cc0443e3ede4d3a4a48d96a843054728a74d4

    mutable value = true;
    for (q in qs) {
        if ( Measure([pauli], [q]) == One ) {
            set value = false;
        }
    }
    return value;
}
```

The Q# language allows dependencies of classical control flow on measurement results of qubits. This in turn enables to implement powerful probabilistic gadgets that can reduce the computational cost for implementing unitaries. As an example, it is easy to implement so-called *Repeat-Until-Success* in Q# which are probabilistic circuits that have an *expected* low cost in terms of elementary gates, but for which the true cost depends on an actual run and an actual interleaving of various possible branchings. 

To facilitate Repeat-Until-Success (RUS) patterns, Q# supports the construct
```qsharp
repeat {
    statementBlock1 
}
until (expression)
fixup {
    statementBlock2
}
```

> where `statementBlock1` and `statementBlock2` are zero or more Q# statements, and `expression` any valid expression that evaluates to a value of type `Bool`. 

Reviewer: ```suggestion
where `statementBlock1` and `statementBlock2` are zero or more Q# statements, and where `expression` is any valid expression that evaluates to a value of type `Bool`. 
```

In a typical use case, the following circuit implements a rotation around an irrational axis of $(I + 2i Z)/\sqrt{5}$ on the Bloch sphere. This is accomplished by using a known RUS pattern: 

```qsharp

> operation RUScircuit (qubit : Qubit) : Unit {

Reviewer: ```suggestion
operation RUSCircuit (qubit : Qubit) : Unit {
```

>     using(ancillas = Qubit[2]) {

Reviewer: ```suggestion
    using (auxillaryQubits = Qubit[2]) {
```

        ApplyToEachA(H, ancillas);
        mutable finished = false;
        repeat {
            Controlled X(ancillas, qubit);
            S(qubit);
            Controlled X(ancillas, qubit);
            Z(qubit);
        }
        until(finished)
        fixup {
            if AllMeasurementsZero(ancillas, Xpauli) {
                set finished = true;
            }
        }
    }
}
```

This example shows the use of a mutable variable `finished` which is in scope of the entire repeat-until-fixup loop and which gets initialized before the loop and updated in the fixup step.

Finally, we show an example of a RUS pattern to prepare a quantum state $\frac{1}{\sqrt{3}}\left(\sqrt{2}\ket{0}+\ket{1}\right)$, starting from the $\ket{+}$ state. See also the [unit testing sample provided with the standard library](https://github.com/Microsoft/Quantum/blob/master/Samples/src/UnitTesting/RepeatUntilSuccessCircuits.qs): 

```qsharp
operation RepeatUntilSuccessStatePreparation( target : Qubit ) : Unit {

>     using( ancilla = Qubit() ) {

Reviewer: ```suggestion
    using (auxillaryQubit = Qubit()) {
```
